src/uucore/src/lib/features/format/num_parser.rs

```diff
@@ -0,0 +1,378 @@
+///
+/// Internally, in order to get the maximum possible precision and cover the full
+/// range of u64 and i64 without losing precision for f64, the returned number is
+/// decomposed into:
+///   - A `base` value
+///   - A `neg` sign bit
+///   - A `integral` positive part
+///   - A `fractional` positive part
+///   - A `precision` representing the number of digits in the fractional part
+///
+/// If the fractional part cannot be represented on a `u64`, parsing continues
+/// silently by ignoring non-significant digits.
+pub struct ParsedNumber {
+    base: Base,
+    negative: bool,
+    integral: u64,
+    fractional: u64,
+    precision: usize,
+}
+
+impl ParsedNumber {
+    fn into_i64(self) -> Option<i64> {
+        if self.negative {
+            i64::try_from(-i128::from(self.integral)).ok()
+        } else {
+            i64::try_from(self.integral).ok()
+        }
+    }
+
+    /// Parse a number as i64. No fractional part is allowed.
+    pub fn parse_i64(input: &str) -> Result<i64, ParseError<'_, i64>> {
+        match Self::parse(input, true) {
+            Ok(v) => v.into_i64().ok_or(ParseError::Overflow),
+            Err(e) => Err(e.map(|v, rest| {
+                v.into_i64()
+                    .map(|v| ParseError::PartialMatch(v, rest))
+                    .unwrap_or(ParseError::Overflow)
+            })),
+        }
+    }
+
+    /// Parse a number as u64. No fractional part is allowed.
+    pub fn parse_u64(input: &str) -> Result<u64, ParseError<'_, u64>> {
+        match Self::parse(input, true) {
+            Ok(v) | Err(ParseError::PartialMatch(v, _)) if v.negative => {
+                Err(ParseError::NotNumeric)
+            }
+            Ok(v) => Ok(v.integral),
+            Err(e) => Err(e.map(|v, rest| ParseError::PartialMatch(v.integral, rest))),
+        }
+    }
+
+    fn into_f64(self) -> f64 {
+        let n = self.integral as f64
+            + (self.fractional as f64) / (self.base as u8 as f64).powf(self.precision as f64);
+        if self.negative {
+            -n
+        } else {
+            n
+        }
+    }
+
+    /// Parse a number as f64
+    pub fn parse_f64(input: &str) -> Result<f64, ParseError<'_, f64>> {
+        match Self::parse(input, false) {
+            Ok(v) => Ok(v.into_f64()),
+            Err(ParseError::NotNumeric) => Self::parse_f64_special_values(input),
+            Err(e) => Err(e.map(|v, rest| ParseError::PartialMatch(v.into_f64(), rest))),
+        }
+    }
+
+    fn parse_f64_special_values(input: &str) -> Result<f64, ParseError<'_, f64>> {
+        let (sign, rest) = if let Some(input) = input.strip_prefix('-') {
+            (-1.0, input)
+        } else {
+            (1.0, input)
+        };
+        let prefix = rest
+            .chars()
+            .take(3)
+            .map(|c| c.to_ascii_lowercase())
+            .collect::<String>();
+        let special = match prefix.as_str() {
+            "inf" => f64::NEG_INFINITY,
```

Comment:
Why is this `NEG_INFINITY`? Wouldn't `INFINITY` make more sense? It seems to work fine because of the `copysign`, though.

Reply:
This was an oversight. Fixed.